File: src/main/scala/com/fabric/spark/DataTransformer.scala
package com.fabric.spark

import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._

/**
 * Utility class for data transformation operations
 */
object DataTransformer {

  /**
   * Clean and standardize column names
   */
  def cleanColumnNames(df: DataFrame): DataFrame = {
    val cleanedColumns = df.columns.map(col => 
      col.toLowerCase.replaceAll("[^a-zA-Z0-9_]", "_")
    )
    
    df.columns.zip(cleanedColumns).foldLeft(df) { case (currentDf, (oldName, newName)) =>
      if (oldName != newName) currentDf.withColumnRenamed(oldName, newName)
      else currentDf
    }
  }

  /**
   * Add metadata columns for tracking
   */
  def addMetadataColumns(df: DataFrame, sourceSystem: String): DataFrame = {
    df.withColumn("source_system", lit(sourceSystem))
      .withColumn("ingestion_timestamp", current_timestamp())
      .withColumn("ingestion_date", current_date())
  }

  /**
   * Filter data based on date range
   */
  def filterByDateRange(df: DataFrame, dateColumn: String, startDate: String, endDate: String): DataFrame = {
    df.filter(col(dateColumn).between(startDate, endDate))
  }

  /**
   * Deduplicate data based on key columns
   */
  def deduplicateData(df: DataFrame, keyColumns: Seq[String]): DataFrame = {
    df.dropDuplicates(keyColumns)
  }

  /**
   * Aggregate data by specified columns
   */
  def aggregateData(df: DataFrame, groupByColumns: Seq[String], aggColumn: String): DataFrame = {
    df.groupBy(groupByColumns.map(col): _*)
      .agg(
        count("*").as("record_count"),
        sum(aggColumn).as(s"total_${aggColumn}"),
        avg(aggColumn).as(s"avg_${aggColumn}"),
        min(aggColumn).as(s"min_${aggColumn}"),
        max(aggColumn).as(s"max_${aggColumn}")
      )
  }
}

File: src/main/python/fabric_spark/data_transformer.py
"""
Data transformation utilities for Fabric Spark applications
"""

from pyspark.sql import DataFrame
from pyspark.sql.functions import *
from typing import List


class DataTransformer:
    """Utility class for data transformation operations"""
    
    @staticmethod
    def clean_column_names(df: DataFrame) -> DataFrame:
        """Clean and standardize column names"""
        import re
        
        new_columns = []
        for col_name in df.columns:
            # Convert to lowercase and replace special chars with underscore
            clean_name = re.sub(r'[^a-zA-Z0-9_]', '_', col_name.lower())
            new_columns.append(clean_name)
        
        # Rename all columns
        for old_name, new_name in zip(df.columns, new_columns):
            if old_name != new_name:
                df = df.withColumnRenamed(old_name, new_name)
        
        return df
    
    @staticmethod
    def add_metadata_columns(df: DataFrame, source_system: str) -> DataFrame:
        """Add metadata columns for tracking"""
        return df \
            .withColumn("source_system", lit(source_system)) \
            .withColumn("ingestion_timestamp", current_timestamp()) \
            .withColumn("ingestion_date", current_date())
    
    @staticmethod
    def filter_by_date_range(df: DataFrame, date_column: str, 
                            start_date: str, end_date: str) -> DataFrame:
        """Filter data based on date range"""
        return df.filter(col(date_column).between(start_date, end_date))
    
    @staticmethod
    def deduplicate_data(df: DataFrame, key_columns: List[str]) -> DataFrame:
        """Deduplicate data based on key columns"""
        return df.dropDuplicates(key_columns)
    
    @staticmethod
    def aggregate_data(df: DataFrame, group_by_columns: List[str], 
                      agg_column: str) -> DataFrame:
        """Aggregate data by specified columns"""
        return df.groupBy(*group_by_columns).agg(
            count("*").alias("record_count"),
            sum(agg_column).alias(f"total_{agg_column}"),
            avg(agg_column).alias(f"avg_{agg_column}"),
            min(agg_column).alias(f"min_{agg_column}"),
            max(agg_column).alias(f"max_{agg_column}")
        )
